tests/src/test/java/alluxio/collections/IndexedSetConcurrencyTest.java

```diff
@@ -0,0 +1,276 @@
+      mSize = ThreadLocalRandom.current().nextInt(0, MAX_SIZE);
+    }
+
+    public long getId() {
+      return mId;
+    }
+
+    public int getSize() {
+      return mSize;
+    }
+  }
+
+  private final IndexedSet.FieldIndex<TestInfo> mIdIndex =
+      new IndexedSet.FieldIndex<TestInfo>() {
+        @Override
+        public Object getFieldValue(TestInfo o) {
+          return o.getId();
+        }
+      };
+
+  private final IndexedSet.FieldIndex<TestInfo> mSizeIndex =
+      new IndexedSet.FieldIndex<TestInfo>() {
+        @Override
+        public Object getFieldValue(TestInfo o) {
+          return o.getSize();
+        }
+      };
+
+  @Before
+  public void before() throws Exception {
+    mIndexedSet = new IndexedSet<>(mIdIndex, mSizeIndex);
+    mThreadPool = Executors.newCachedThreadPool();
+    mStopThreads = new AtomicBoolean(false);
+  }
+
+  @After
+  public void after() {
+    mThreadPool.shutdownNow();
+  }
+
+  /**
+   * Verifies the {@link #mIndexedSet} for internal consistency.
+   */
+  private void verifySet() {
+    Iterator<TestInfo> it = mIndexedSet.iterator();
+    Set<Long> ids = new HashSet<>();
+    Set<Integer> sizes = new HashSet<>();
+
+    // Verify the size.
+    int expectedCount = 0;
+    while (it.hasNext()) {
+      TestInfo info = it.next();
+      ids.add(info.getId());
+      sizes.add(info.getSize());
+      expectedCount++;
+    }
+    Assert.assertEquals(expectedCount, mIndexedSet.size());
+
+    // Verify the size according to the id index.
+    int count = 0;
+    for (Long id : ids) {
+      Set<TestInfo> elements = mIndexedSet.getByField(mIdIndex, id);
+      count += elements.size();
+    }
+    Assert.assertEquals(expectedCount, count);
+
+    // Verify the size according to the size index.
+    count = 0;
+    for (Integer size : sizes) {
+      Set<TestInfo> elements = mIndexedSet.getByField(mSizeIndex, size);
+      count += elements.size();
+    }
+    Assert.assertEquals(expectedCount, count);
+  }
+
+  @Test
+  public void basicConcurrentUpdateTest() throws Exception {
+    List<Future<?>> futures = new ArrayList<>();
+    List<ConcurrentTask> addTasks = new ArrayList<>();
+    List<ConcurrentTask> removeTasks = new ArrayList<>();
+
+    // Add random number of each task type.
+    for (int i = 2 * ThreadLocalRandom.current().nextInt(MIN_TASKS, MAX_TASKS + 1); i > 0; i--) {
+      // Try to balance adds and removes
+      addTasks.add(new ConcurrentAdd());
+    }
+    for (int i = ThreadLocalRandom.current().nextInt(MIN_TASKS, MAX_TASKS + 1); i > 0; i--) {
+      removeTasks.add(new ConcurrentRemove());
+    }
+    for (int i = ThreadLocalRandom.current().nextInt(MIN_TASKS, MAX_TASKS + 1); i > 0; i--) {
+      removeTasks.add(new ConcurrentRemoveByField());
+    }
+
+    for (ConcurrentTask task : addTasks) {
+      futures.add(mThreadPool.submit(task));
+    }
+    for (ConcurrentTask task : removeTasks) {
+      futures.add(mThreadPool.submit(task));
```

Comment:
`invokeAll()` seems to wait for all threads to complete, which does not allow the main thread to signal the threads to stop.